std/src/main/scala/algebra/std/Util.scala

```diff
@@ -0,0 +1,163 @@
+
+    def gcd0(val0: Int, exp0: Int, val1: Int, exp1: Int): Float = {
+      val tz0 = numberOfTrailingZeros(val0)
+      val tz1 = numberOfTrailingZeros(val1)
+      val tzShared = Math.min(tz0, tz1 + exp1 - exp0)
+      val n = gcd(val0.toLong >>> tz0, val1.toLong >>> tz1).toInt << tzShared
+
+      val shift = numberOfLeadingZeros(n) - 8 // Number of bits to move 1 to bit 23
+      val mantissa = (n << shift) & 0x007FFFFF
+      val exp = (exp0 - shift)
+      if (exp < 0) 0F else intBitsToFloat((exp << 23) | mantissa)
+    }
+
+    if (a == 0F) b
+    else if (b == 0F) a
+    else {
+      val aBits = floatToIntBits(a)
+      val aVal = value(aBits)
+      val aExp = exp(aBits)
+
+      val bBits = floatToIntBits(b)
+      val bVal = value(bBits)
+      val bExp = exp(bBits)
+
+      if (aExp < bExp) gcd0(aVal, aExp, bVal, bExp)
+      else gcd0(bVal, bExp, aVal, aExp)
+    }
+  }
+
+  /**
+   * GCD for Double values.
+   */
+  final def gcd(a: Double, b: Double): Double = {
+    def value(bits: Long): Long = bits & 0x000FFFFFFFFFFFFFL | 0x0010000000000000L
+
+    def exp(bits: Long): Int = ((bits >> 52) & 0x7FF).toInt
+
+    def gcd0(val0: Long, exp0: Int, val1: Long, exp1: Int): Double = {
+      val tz0 = numberOfTrailingZeros(val0)
+      val tz1 = numberOfTrailingZeros(val1)
+      val tzShared = Math.min(tz0, tz1 + exp1 - exp0)
+      val n = gcd(val0 >>> tz0, val1 >>> tz1) << tzShared
+
+      val shift = numberOfLeadingZeros(n) - 11 // Number of bits to move 1 to bit 52
+      val mantissa = (n << shift) & 0x000FFFFFFFFFFFFFL
+      val exp = (exp0 - shift).toLong
+      if (exp < 0) 0.0 else longBitsToDouble((exp << 52) | mantissa)
+    }
+
+    if (a == 0D) b
+    else if (b == 0D) a
+    else {
+      val aBits = doubleToLongBits(a)
+      val aVal = value(aBits)
+      val aExp = exp(aBits)
+
+      val bBits = doubleToLongBits(b)
+      val bVal = value(bBits)
+      val bExp = exp(bBits)
+
+      if (aExp < bExp) gcd0(aVal, aExp, bVal, bExp)
+      else gcd0(bVal, bExp, aVal, aExp)
+    }
+  }
+
+  /**
+   * Exponentiation function, e.g. x^y
+   *
+   * If base^ex doesn't fit in a Long, the result will overflow (unlike
+   * Math.pow which will return +/- Infinity).
+   */
+  final def pow(base: Long, exponent: Long): Long = {
+    @tailrec def loop(t: Long, b: Long, e: Long): Long =
+      if (e == 0L) t
+      else if ((e & 1) == 1) loop(t * b, b * b, e >>> 1L)
+      else loop(t, b * b, e >>> 1L)
+
+    if (exponent >= 0L) loop(1L, base, exponent) else {
+      if(base == 0L) throw new ArithmeticException("zero can't be raised to negative power")
+      else if (base == 1L) 1L
+      else if (base == -1L) if ((exponent & 1L) == 0L) -1L else 1L
+      else 0L
+    }
+  }
+
+  def combineMapCommutative[K, V](x: Map[K, V], y: Map[K, V])(f: (V, V) => V): Map[K, V] =
+    if (y.size < x.size) combineMapCommutative(y, x)(f) else {
+      x.foldLeft(y) { case (z, (k, v1)) =>
```

Comment:
we found significant performance wins in creating a mutable map internally and then wrapping returning an immutable wrapper view of it (where we copy to an immutable map only on +/-). Since mostly in scalding we are aggregating these things, we just call + repeatedly.

Also not, sumOption was basically born out of this case because it is much better to allocate one mutable map and go through mutating it, then return an immutable wrapper, than to keep using the immutable.HashMap.